### ISSUE TITLE
Include final softmax in forward() to run on GPU when compiling w/ aoti or other accelerator on et

### ISSUE THREAD

**Issue by sunshinesfbay** [pytorch/torchchat#1012]
### 🚀 The feature, motivation and pitch

At present, the softmax that helps pick the next token is seperately implemented and executed in the surrounding binary / runner (e.g., the included runner).  It would be great to (optionally?) make it part for forward() so it may execute a more optimized kernel and on an accelerator, when one is available?

This would likely require the temperature code and all to move into the pytorch model as well, so might require some surgery.

### Alternatives

The current implementation works, and for GPU we could call a GPU softmax implementation from runner. (That will be much messier.  And not clear how it would apply to on-device.)

**Comment by larryliu0820** [comment]
Hi @sunshinesfbay thanks for the feedback! I can provide more context on how run.cpp is written.

> It would be great to (optionally?) make it part for forward()

From llama2/3 perspective, leaving the softmax out of the model is on purpose, so that the user can dynamically choose whether they want to sample the logits (where softmax is needed) or simply do a argmax (no softmax) based on temperature. Please see the official python [code](https://github.com/meta-llama/llama3/blob/main/llama/generation.py#L181-L185). As you mentioned, having this softmax inside the model will require temperature as part of the forward() input, and that significantly changes the model semantics.

>  so it may execute a more optimized kernel and on an accelerator, when one is available?

We provide `run.cpp` as a minimal example to demonstrate the e2e workflow on AOTI and ExecuTorch. In case you haven't noticed, we borrowed a lot of code from [llama2.c](https://github.com/karpathy/llama2.c). We encourage users like you to create your own fork of the code and modify as you see fit. 

Let us know if this answers your question!

**Comment by sunshinesfbay** [comment]
Thanks.
